testament/testament.nim

```diff
@@ -1220,155 +1497,908 @@ proc parseArgs(execState: var Execution, p: var OptParser): ParseCliResult =
+      if isCompilerRepo: # handle `AdditionalCategories`
+        for cat in AdditionalCategories:
+          testFilesFromCat(execState, Category(cat))
+
+    of tfkCategories:
+      for cat in filter.cats:
+        testFilesFromCat(execState, cat)
+
+    of tfkGlob:
+      execState.categories = @[Category "<glob>"]
+      let pattern = filter.pattern
+      if dirExists(pattern):
+        for kind, name in walkDir(pattern):
+          if kind in {pcFile, pcLinkToFile} and name.endsWith(".nim"):
+            execState.testFiles.add TestFile(file: name)
+      else:
+        for name in walkPattern(pattern):
+          execState.testFiles.add TestFile(file: name)
+
+    of tfkSingle:
+      execState.categories = @[Category parentDir(filter.test)]
+      let test = filter.test
+      # IMPLEMENT: replace with proper error handling
+      doAssert fileExists(test), test & " test does not exist"
+      if isTestFile(test):
+        execState.testFiles.add TestFile(file: test)
+
+    else:
+      assert false, "TODO ???"
+
+  execState.testFiles.sort # ensures we have a reproducible ordering
+
+  # parse all specs
+  for testId, test in pairs(execState.testFiles):
+    execState.testSpecs.add parseSpec(
+      addFileExt(test.file, ".nim"),
+      execState.categories[test.catId].defaultTargets(),
+      nativeTarget()
+    )
+
+    if execState.testOpts.hasKey(testId):
+      # apply additional test matrix, if specified
+      let optMatrix = execState.testOpts[testId].optMatrix
+      execState.testSpecs[testId].matrix =
+        case execState.testSpecs[testId].matrix.len
+        of 0:
+          optMatrix
+        else:
+          # (saem)REFACTOR - this is a hack, we shouldn't modify the
+          # spec.matrix
+
+          # (haxscramper)QUESTION Doesn't "specify additional test matrix
+          # parameter" imply modification of the test matrix? It seems like
+          # the most direct way to implement this feature: additional test
+          # matrix parameter is ... added to the test matrix, seems pretty
+          # logical to me.
+          var tmp: seq[string] = @[]
+          for o in optMatrix.items:
+            for e in execState.testSpecs[testId].matrix.items:
+              # REFACTOR Switch to `seq[string]` for matrix flags
+              tmp.add o & " " & e
+          tmp
+
+      # apply action override, if specified
+      let actionOverride = execState.testOpts[testId].action
+      if actionOverride.isSome:
+        execState.testSpecs[testId].action = actionOverride.get
+
+proc makeName(test: TestFile,
+              testRun: TestRun,
+              allowFailure: bool
+              ): string =
+  let
+    target = testRun.target
+    matrixEntry = testRun.matrixEntry
+  result = test.file.changeFileExt("").replace(DirSep, '/')
+  result.add '_' & $target
+  if matrixEntry != noMatrixEntry:
+    result.add "[$1]" % $matrixEntry
+  if allowFailure:
+    result.add " (allowed to fail) "
+  # if test.options.len > 0:
```

Comment:
this maybe an issue wrt to test name uniqueness (parallelism).

Reply:
Weird, don't remember removing it; options certainly need to be added as a part of name